pkg/engine/artifacts.go

```diff
@@ -13,258 +13,258 @@ import (
 	"github.com/Azure/aks-engine/pkg/api/common"
 )
 
-type kubernetesFeatureSetting struct {
-	sourceFile      string
-	destinationFile string
-	isEnabled       bool
-	rawScript       string
+type kubernetesComponentFileSpec struct {
+	sourceFile      string // filename to source spec data from
+	base64Data      string // if not "", this base64-encoded string will take precedent over sourceFile
+	destinationFile string // the filename to write to disk on the destination OS
+	isEnabled       bool   // is this spec enabled?
 }
 
-func kubernetesContainerAddonSettingsInit(profile *api.Properties) map[string]kubernetesFeatureSetting {
-	return map[string]kubernetesFeatureSetting{
+func kubernetesContainerAddonSettingsInit(profile *api.Properties) map[string]kubernetesComponentFileSpec {
+	return map[string]kubernetesComponentFileSpec{
 		DefaultHeapsterAddonName: {
-			"kubernetesmasteraddons-heapster-deployment.yaml",
-			"kube-heapster-deployment.yaml",
-			!common.IsKubernetesVersionGe(profile.OrchestratorProfile.OrchestratorVersion, "1.13.0"),
-			profile.OrchestratorProfile.KubernetesConfig.GetAddonScript(DefaultKubeHeapsterDeploymentAddonName),
-		},
-		DefaultMetricsServerAddonName: {
-			"kubernetesmasteraddons-metrics-server-deployment.yaml",
-			"kube-metrics-server-deployment.yaml",
-			profile.OrchestratorProfile.IsMetricsServerEnabled(),
-			profile.OrchestratorProfile.KubernetesConfig.GetAddonScript(DefaultMetricsServerAddonName),
-		},
-		DefaultTillerAddonName: {
-			"kubernetesmasteraddons-tiller-deployment.yaml",
-			"kube-tiller-deployment.yaml",
-			profile.OrchestratorProfile.KubernetesConfig.IsTillerEnabled(),
-			profile.OrchestratorProfile.KubernetesConfig.GetAddonScript(DefaultTillerAddonName),
-		},
-		DefaultAADPodIdentityAddonName: {
-			"kubernetesmasteraddons-aad-pod-identity-deployment.yaml",
-			"aad-pod-identity-deployment.yaml",
-			profile.IsAADPodIdentityEnabled(),
-			profile.OrchestratorProfile.KubernetesConfig.GetAddonScript(DefaultAADPodIdentityAddonName),
-		},
-		DefaultACIConnectorAddonName: {
-			"kubernetesmasteraddons-aci-connector-deployment.yaml",
-			"aci-connector-deployment.yaml",
-			profile.IsACIConnectorEnabled(),
-			profile.OrchestratorProfile.KubernetesConfig.GetAddonScript(DefaultACIConnectorAddonName),
-		},
-		DefaultClusterAutoscalerAddonName: {
-			"kubernetesmasteraddons-cluster-autoscaler-deployment.yaml",
-			"cluster-autoscaler-deployment.yaml",
-			profile.IsClusterAutoscalerEnabled(),
-			profile.OrchestratorProfile.KubernetesConfig.GetAddonScript(DefaultClusterAutoscalerAddonName),
-		},
-		DefaultBlobfuseFlexVolumeAddonName: {
-			"kubernetesmasteraddons-blobfuse-flexvolume-installer.yaml",
-			"blobfuse-flexvolume-installer.yaml",
-			profile.IsBlobfuseFlexVolumeEnabled() && !profile.HasCoreOS(),
-			profile.OrchestratorProfile.KubernetesConfig.GetAddonScript(DefaultBlobfuseFlexVolumeAddonName),
+			sourceFile:      "kubernetesmasteraddons-heapster-deployment.yaml",
```

Comment:
using a more expressive literal struct implementation where we actually declare the key that each value represents